### PR TITLE
Add cereal support

### PR DESCRIPTION
**Pull request by eseiler**
This PR adds cereal support to the SDSL.

The `test-sdsl` test suite passes for both cereal and SDSL serialisation.

Todo:

- [x] Add cereal as submodule
- [ ] pin cereal version? (currently pinned against latest release)
- [x] docs
- [x] tests (currently without cereal)
- [x] we need two build targets for the tests (one with cereal `-DSDSL_CEREAL=1` , one without)
- [x] currently, I removed some downloaded test files since they are not readable by cereal, but we could still add them for non-cereal tests
- [x] code formatting

@h-2 @cpockrandt feel free to already have a look

### CODE REVIEW

**Comment by eseiler** [comment]
Apparently this PR doesn't break the Non-Cereal version.

How do we want to handle the cereal version? Do **one** run in CI with cereal on?

**Comment by cpockrandt** [comment]
Yes, definitely! Debug builds take quite long and I don't see a reason to run it in debug mode. So maybe we add a release build with cereal to Travis and Jenkins?

**Comment by eseiler** [comment]
Ok. Can I just edit the files in the PR and it will update the CI for this PR?

**Comment by cpockrandt** [comment]
Yes :)

File: include/sdsl/bit_vector_il.hpp [hunk 1]
```diff
@@ -312,6 +334,18 @@ class rank_support_il {
 	{
 		return serialize_empty_object(out, v, name, this);
 	}
+
+	template <typename archive_t>
+	void CEREAL_SAVE_FUNCTION_NAME(archive_t & ar) const
+	{
+		(void) ar;
```
Comment (cpockrandt) [inline include/sdsl/bit_vector_il.hpp]: Can you remove the variable name instead of casting it to void?

**Comment by eseiler** [comment]
@cpockrandt This is good to merge from my POV.

**Comment by h-2** [comment]
> @cpockrandt This is good to merge from my POV.

👍 🎉

File: include/sdsl/lcp.hpp [hunk 2]
```diff
@@ -156,6 +156,38 @@ void load_lcp(t_lcp& lcp, std::istream& in, const t_cst& cst, lcp_tree_and_lf_co
...
+void set_lcp_pointer(t_lcp& lcp, const t_cst& cst)
+{
+	typename t_lcp::lcp_category tag;
+	set_lcp_pointer(lcp, cst, tag);
+}
+
+template <class t_lcp, class t_cst>
+void set_lcp_pointer(t_lcp& lcp, const t_cst&, lcp_plain_tag)
+{
+	(void) lcp;
```
Comment (cpockrandt) [inline include/sdsl/lcp.hpp]: Remove void-cast and parameter name :)

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -14,11 +14,11 @@ function(generate_target_for_special_inputs tc_name abs_tc_name)
         if ("${tc_name}" MATCHES "${int_vec_regex}")
             string(REPLACE "int-vec." "" tc_suf ${tc_name})
             string(REPLACE "." ";" tc_param ${tc_suf})  #  insert semicolons to get a list :)
-        else()
```
Comment (cpockrandt) [inline test/CMakeLists.txt]: Can you remove all whitespace changes in this file?

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -63,6 +63,10 @@ include_directories(${CMAKE_CURRENT_BINARY_DIR}/../include
 
 link_libraries(gtest)
 
+if(SDSL_CEREAL)
+include_directories(${CMAKE_CURRENT_BINARY_DIR}/../external/cereal/include)
```
Comment (cpockrandt) [inline test/CMakeLists.txt]: Indention :)

File: include/sdsl/construct_lcp.hpp [hunk 5]
```diff
@@ -28,7 +28,7 @@
 
 namespace sdsl {
 
-// Forward declaration of construction method
```
Comment (cpockrandt) [inline include/sdsl/construct_lcp.hpp]: whitespace change!

File: include/sdsl/int_vector_buffer.hpp [hunk 6]
```diff
@@ -489,7 +489,6 @@ class int_vector_buffer {
 		inline difference_type operator-(const iterator& it) { return (m_idx - it.m_idx); }
 	};
 };
-
```
Comment (cpockrandt) [inline include/sdsl/int_vector_buffer.hpp]: remove file from commit pls :)

File: include/sdsl/int_vector_mapper.hpp [hunk 7]
```diff
@@ -63,10 +63,10 @@ class int_vector_mapper {
 						//    out.seekp(0, std::ios::end);
 					} else {
                         std::cerr << "int_vector_mapper: could not open file for header update" << std::endl;
-/*                        
+/*
```
Comment (cpockrandt) [inline include/sdsl/int_vector_mapper.hpp]: remove whitespace changes in this file

File: include/sdsl/io.hpp [hunk 8]
```diff
@@ -384,9 +385,18 @@ uint64_t serialize_vector(const std::vector<T>&		 vec,
 template <typename T>
 void load_vector(std::vector<T>& vec, std::istream& in)
 {
+#if SDSL_HAS_CEREAL
+	{
```
Comment (cpockrandt) [inline include/sdsl/io.hpp]: Is there a reason why you scope serialization everywhere in this file?

File: include/sdsl/wm_int.hpp [hunk 9]
```diff
@@ -116,6 +116,57 @@ class wm_int {
 			int_vector<> rac(m_size, 0, m_max_level);
 			std::copy(begin, end, rac.begin());
 
+#if SDSL_HAS_CEREAL
+			// buffer for elements in the right node
```
Comment (cpockrandt) [inline include/sdsl/wm_int.hpp]: This looks like a huge code duplication :open_mouth:

File: include/sdsl/wt_int.hpp [hunk 10]
```diff
@@ -173,6 +173,61 @@ class wt_int {
 			std::copy(begin, end, rac.begin());
 
 			// buffer for elements in the right node
+#if SDSL_HAS_CEREAL
+			int_vector<> buf1;
```
Comment (cpockrandt) [inline include/sdsl/wt_int.hpp]: Code duplication :open_mouth:

File: test/k2_tree_test.cpp [hunk 11]
```diff
@@ -42,9 +44,10 @@ template<typename t_tree>
 void check_serialize_load(t_tree& tree)
 {
     auto unserialized_tree = t_tree();
-    std::stringstream ss;
-    tree.serialize(ss);
-    unserialized_tree.load(ss);
+    std::string file_name = temp_dir+"/int_vector";
+    // std::stringstream ss;
```
Comment (cpockrandt) [inline test/k2_tree_test.cpp]: I think we remove the comments

File: include/sdsl/io.hpp (hunk 8 above)
Comment (eseiler) [reply include/sdsl/io.hpp]: To ensure the archive is flushed.

Comment (cpockrandt) [reply include/sdsl/io.hpp]: But it should be flushed by the end of the function in the next line anyway?

**Comment by h-2** [comment]
> This might take one or two days.

🐌......

File: include/sdsl/io.hpp (hunk 8 above)
Comment (eseiler) [reply include/sdsl/io.hpp]: Idk, cereal recommends doing it this way since it's the only way to ensure it.

**Comment by eseiler** [comment]
I think I removed all the whitespace changes...

File: include/sdsl/wm_int.hpp (hunk 9 above)
Comment (eseiler) [reply include/sdsl/wm_int.hpp]: I refactored it

File: include/sdsl/wt_int.hpp (hunk 10 above)
Comment (eseiler) [reply include/sdsl/wt_int.hpp]: I refactored it

File: include/sdsl/io.hpp (hunk 8 above)
Comment (cpockrandt) [reply include/sdsl/io.hpp]: > Some archives in cereal can only safely finish flushing their contents upon their destruction.

Whether it is destructed by an additional scope or by the end of the function in the next line does not matter. I don't see any benefit here to add an additional scope.

File: test/sa_construct_test.cpp [hunk 12]
```diff
@@ -40,7 +40,12 @@ TEST_F(sa_construct_test, divsufsort)
     construct_sa<8>(config);
     memory_monitor::stop();
     {
+        #if SDSL_HAS_CEREAL
```
Comment (cpockrandt) [inline test/sa_construct_test.cpp]: Please remove the indention for the `if` in this file as you did in the other files.

File: test/wt_int_test.cpp [hunk 13]
```diff
@@ -77,12 +77,22 @@ TYPED_TEST(wt_int_test, constructor)
         ASSERT_TRUE(store_to_file(wt, temp_file));
     }
     {
+        #if SDSL_HAS_CEREAL
```
Comment (cpockrandt) [inline test/wt_int_test.cpp]: Please remove the indention for the `if` in this file as you did in the other files.

File: include/sdsl/int_vector.hpp [hunk 14]
```diff
@@ -648,6 +648,26 @@ class int_vector {
 	//! Load the int_vector for a stream.
 	void load(std::istream& in);
 
+	// //!\brief Serialise (save) via cereal if archive is not binary
```
Comment (cpockrandt) [inline include/sdsl/int_vector.hpp]: Can you make a small comment and explain why different code is necessary? maybe a reference to the `std::vector` serialization of cereal?

File: external/CMakeLists.txt [hunk 15]
```diff
@@ -15,4 +15,23 @@ endif(NOT EXISTS ${gtest_file})
...
+## Add cereal
+if(SDSL_CEREAL)
+	MESSAGE(STATUS "Adding cereal")
+	set(cereal_dir ${CMAKE_CURRENT_LIST_DIR}/cereal/include/cereal)
+	set(cereal_file ${cereal_dir}/cereal.hpp)
+	if(NOT EXISTS ${cereal_file})
+	    EXECUTE_PROCESS(COMMAND git submodule update --init external/cereal
+	                    WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}/..
+	                    OUTPUT_QUIET
+	            )
```
[the remark this answers is not in the document]
Comment (eseiler) [reply external/CMakeLists.txt]: I copy pasted from 14 lines above

File: include/sdsl/cereal.hpp [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+#ifndef INCLUDED_SDSL_CEREAL
+#define INCLUDED_SDSL_CEREAL
+
+#if defined(__has_include)
+	#if __has_include(<cereal/cereal.hpp>)
+		#define SDSL_HAS_CEREAL 1
+		#include <cereal/cereal.hpp>
+		#include <cereal/details/traits.hpp>
+		#include <cereal/archives/binary.hpp>
+		#include <cereal/archives/json.hpp>
```
[the remark this answers is not in the document]
Comment (eseiler) [reply include/sdsl/cereal.hpp]: Mhm, I guess I either include them here, or add them to each unit test.

File: include/sdsl/csa_sada.hpp [hunk 17]
```diff
@@ -462,5 +493,38 @@ std::istream& in)
 	m_alphabet.load(in);
 }
 
+template <class t_enc_vec,
+		  uint32_t t_dens,
```
[the remark this answers is not in the document]
Comment (eseiler) [reply include/sdsl/csa_sada.hpp]: I copy/pasted from the function above.
I'd like to keep the original formatting.

File: include/sdsl/int_vector.hpp (hunk 14 above)
Comment (eseiler) [reply include/sdsl/int_vector.hpp]: I'll add this once the CI judged the code.

File: include/sdsl/nearest_neighbour_dictionary.hpp [hunk 18]
```diff
@@ -242,6 +242,44 @@ class nearest_neighbour_dictionary {
...
+		ar(CEREAL_NVP(m_differences));
+		ar(CEREAL_NVP(m_contains_abs_sample));
+		ar(CEREAL_NVP(m_rank_contains_abs_sample));
+		m_rank_contains_abs_sample.set_vector(&m_contains_abs_sample);
+	}
+
+	//! Equality operator.
+	bool operator==(nearest_neighbour_dictionary const &) const noexcept
+	{
+		return true; /*(m_ones == other.m_ones) && (m_size == other.m_size) &&
```
Comment (cpockrandt) [inline include/sdsl/nearest_neighbour_dictionary.hpp]: Did you forget to uncomment this?

File: include/sdsl/rank_support_scan.hpp [hunk 19]
```diff
@@ -52,6 +52,20 @@ class rank_support_scan : public rank_support {
 	}
 	void load(std::istream&, const int_vector<1>* v = nullptr) { set_vector(v); }
 	void set_vector(const bit_vector* v = nullptr) { m_v = v; }
+	template <typename archive_t>
+	void CEREAL_SAVE_FUNCTION_NAME(archive_t &) const {}
+	template <typename archive_t>
+	void CEREAL_LOAD_FUNCTION_NAME(archive_t &) {}
+	//! Equality operator.
```
Comment (cpockrandt) [inline include/sdsl/rank_support_scan.hpp]:
```suggestion

	//! Equality operator.
```

File: include/sdsl/rank_support_scan.hpp [hunk 20]
```diff
@@ -52,6 +52,20 @@ class rank_support_scan : public rank_support {
...
+	bool operator==(rank_support_scan const & other) const noexcept
+	{
+		return (*m_v == *other.m_v);
+	}
+	//! Inequality operator.
```
Comment (cpockrandt) [inline include/sdsl/rank_support_scan.hpp]:
```suggestion

	//! Inequality operator.
```

File: include/sdsl/int_vector.hpp (hunk 14 above)
Comment (cpockrandt) [reply include/sdsl/int_vector.hpp]: Ok. And I noticed `//` is a duplicate

File: test/k2_tree_test.cpp [hunk 21]
```diff
@@ -538,6 +538,36 @@ TYPED_TEST(k2_tree_test, serialize_test)
...
+		in_archive_t iarchive{ss};
+		iarchive(in_l);
+		EXPECT_EQ(l, in_l);
+	}
+}
+
+TYPED_TEST(k2_tree_test, cereal)
+{
+	TypeParam k2tree;
+        // ASSERT_TRUE(load_from_file(k2treap, temp_file));
```
Comment (cpockrandt) [inline test/k2_tree_test.cpp]: Should this not be commented in?

**Comment by cpockrandt** [comment]
Besides the remaining comments I think we can finally merge it :)

File: include/sdsl/nearest_neighbour_dictionary.hpp (hunk 18 above)
Comment (eseiler) [reply include/sdsl/nearest_neighbour_dictionary.hpp]: Ups 🤦‍♂️

File: test/k2_tree_test.cpp (hunk 21 above)
Comment (eseiler) [reply test/k2_tree_test.cpp]: No, there is no `temp_file` for this test, but I wanted to add a construction.

Comment (cpockrandt) [reply test/k2_tree_test.cpp]: Ok, then let's take a single non-trivial test:

```suggestion
    vector<vector<int>> mat({{1, 0, 0, 0, 1},
        {0, 0, 0, 0, 0},
        {0, 0, 1, 1, 0},
        {0, 0, 0, 0, 0},
        {0, 0, 1, 0, 1}
    });
    k2tree = TypeParam(mat);
```

**Pull request merged**